File: app/src/main/java/com/garymcgowan/moviepedia/view/search/MovieListActivityPresenter.java
package com.garymcgowan.moviepedia.view.search;

import com.garymcgowan.moviepedia.model.Movie;
import com.garymcgowan.moviepedia.model.MovieRepository;
import com.garymcgowan.moviepedia.view.BasePresenter;

import java.util.List;

import io.reactivex.Flowable;
import io.reactivex.Scheduler;
import io.reactivex.annotations.NonNull;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.subscribers.DisposableSubscriber;
import timber.log.Timber;

/**
 * Created by garymcgowan on 08/05/2017.
 */

public class MovieListActivityPresenter extends BasePresenter {

    private static final long QUERY_UPDATE_DELAY_MILLIS = 400;

    private MovieListActivityView view;
    private MovieRepository movieRepository;
    private Scheduler mainScheduler;

    public MovieListActivityPresenter(MovieListActivityView view, MovieRepository movieRepository, Scheduler mainScheduler) {
        this.view = view;
        this.movieRepository = movieRepository;
        this.mainScheduler = mainScheduler;
    }

    public void setSearchTermObservable(@NonNull Flowable<String> flowable) {
        addDisposable(flowable
                //.filter(s -> s.length() > 1)
                .subscribeOn(mainScheduler)
                //.debounce(QUERY_UPDATE_DELAY_MILLIS, TimeUnit.MILLISECONDS)
                .flatMap(search -> movieRepository.getMovieSearch(search)
                        .subscribeOn(Schedulers.io())
                )
                .observeOn(mainScheduler)
                .subscribeWith(new DisposableSubscriber<List<Movie>>() {
                    @Override
                    public void onNext(List<Movie> search) {
                        if (search == null)
                            view.displayError("movies null");
                        else
                            view.displayMovies(search);
                    }

                    @Override
                    public void onError(Throwable t) {
                        Timber.d("onError: " + t);
                        Timber.e(t);
                        view.displayError("onError: " + t.getLocalizedMessage());
                    }

                    @Override
                    public void onComplete() {
                    }
                }));
    }
}

File: app/src/main/java/com/garymcgowan/moviepedia/view/search/MovieListActivityPresenter.kt
package com.garymcgowan.moviepedia.view.search

import com.garymcgowan.moviepedia.model.Movie
import com.garymcgowan.moviepedia.model.MovieRepository
import com.garymcgowan.moviepedia.persistence.StoredMovie
import com.garymcgowan.moviepedia.persistence.StoredMovieDao
import com.garymcgowan.moviepedia.view.BasePresenter
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Scheduler
import io.reactivex.annotations.NonNull
import io.reactivex.schedulers.Schedulers
import timber.log.Timber

class MovieListActivityPresenter(
        private val view: MovieListActivityContract.View,
        private val movieRepository: MovieRepository,
        private val storedMovieDao: StoredMovieDao,
        private val mainScheduler: Scheduler) : BasePresenter(), MovieListActivityContract.Presenter {

    companion object {
        private val QUERY_UPDATE_DELAY_MILLIS: Long = 400
    }

    override fun setSearchTermObservable(@NonNull flowable: Flowable<String>) {
        addDisposable(flowable
                //.filter(s -> s.length() > 1)
                .subscribeOn(mainScheduler)
                //.debounce(QUERY_UPDATE_DELAY_MILLIS, TimeUnit.MILLISECONDS)
                .flatMap { search ->
                    movieRepository.getMovieSearch(search).subscribeOn(Schedulers.io())
                }
                .map { list ->
                    list.toMutableList().apply { addAll(storedMovieDao.getAll().map { it.toMovie() }) }
                }
                .observeOn(mainScheduler)
                .subscribe(
                        { search ->
                            if (search == null) view.displayError("movies null")
                            else view.displayMovies(search)
                        },
                        { error ->
                            Timber.d("onError: $error")
                            Timber.e(error)
                            view.displayError("onError: " + error.localizedMessage)
                        }

                ))
    }

    override fun favouriteClicked(fav: Movie, isFav: Boolean) {

        //TODO: imrpove this so the UI is updated
        addDisposable(Completable.fromAction {
            if (isFav)
                storedMovieDao.insertAll(StoredMovie(fav))
            else
                storedMovieDao.deleteAll(StoredMovie(fav))
        }
                .subscribeOn(Schedulers.io())
                .subscribe()
        )
    }


}
